cranelift-codegen/src/peephole.rs

```diff
@@ -0,0 +1,469 @@
+    domtree: &'a DominatorTree,
+}
+
+impl<'a> Context<'a> {
+    fn run(&mut self) {
+        self.cleanup();
+    }
+}
+
+// ======== A description of the redundant-fill-removal algorithm ========
+//
+// The simplest incarnation simply detects fills from the same stack slot into
+// the same register, and removes them:
+//
+//  v1 = fill v0
+//  ...
+//  v2 = fill v0
+//
+// The second fill is redundant provided that (1) v1 and v2 have the same
+// register, (2) that "..." overwrites neither that register nor v0's stack
+// slot, and (3) that the second fill is dominated by the first.  (3) is
+// implied since we're doing this only inside EBBs, but we have to check (1)
+// and (2).
+//
+// In this example, the instruction processing loop below creates an
+// environment entry on encountering "v1 = fill v0", and uses that entry to
+// determine that "v2 = fill v0" is redundant.
+//
+// The loop must also, even in the simplest version, check all instructions
+// for writes to registers and stack slots, and invalidate environment entries
+// accordingly.
+//
+// === First refinement -- implemented
+//
+// This is to observe that spills also create equivalences between registers
+// and stack slot values, which in turn increases our changes of proving later
+// fills redundant.
+//
+// === Second refinement -- implemented
+//
+// For a fill, even if we can't find an exact match in the environment,
+// finding a load of the same spill slot into a different register is still a
+// win, since we can convert the fill into a reg-reg copy.
+//
+// === Other tricks that don't seem to help
+//
+// I tried tracking the effects of InstructionData::Copy in the environment,
+// for the case when both source and destination are registers.  But that
+// didn't cause any more fills to be removed or converted to copies.  I'm not
+// sure why.
+//
+// It's also possible for a spill to be detected as redundant, if such a
+// binding already appears in the environment.  I implemented this, but it
+// seems almost entirely ineffective, so I removed it again.
+
+// This describes a spill slot in the obvious way.  It also tracks the SSA
+// value name associated with the slot.
+#[derive(Clone, Copy)]
+struct SlotInfo {
+    value: Value,
+    offs: i32,
+    size: u32,
+}
+
+// Find out if the range `[offs, +size)` overlaps the with range in `si`.
+fn any_overlap(si: &SlotInfo, offs: i32, size: u32) -> bool {
+    let a_offs = si.offs as i64;
+    let a_size = si.size as i64;
+    let b_offs = offs as i64;
+    let b_size = size as i64;
+    let no_overlap = a_offs + a_size <= b_offs || b_offs + b_size <= a_offs;
+    !no_overlap
+}
+
+// Find, in `banks`, the register bank that `reg` lives in, and return the
+// lower limit and size of the bank.  This is so the caller can conveniently
+// iterate over all `RegUnits` in the bank that `reg` lives in.
+fn find_bank_limits(banks: &[RegBank], reg: RegUnit) -> (RegUnit, u16) {
+    for rb in banks.iter() {
```

Comment:
I'm not fond of loops...  I don't know this data structure well enough, but if we can go from a regunit to a regclass to a top-level regclass to a regbank without looping then that would be good.  Worth investigating a little more deeply, but could be followup work, depending on measured performance impact.

Reply:
I changed it to use `RegInfo::bank_containing_regunit`.  I don't see a way to do it without the (implied) loop.  But as @bnjbvr says, most targets contain only 2 or 3 banks, of which normally the first is the integer registers, and most register references are to integer registers.  So I'd expect almost always to get a hit in the first iteration of the loop.